File: runtime-swift/KalamRuntime.swift
import Foundation
#if canImport(Darwin)
import Darwin
#endif

// MARK: - Constants

private let protocolVersion: UInt8 = 1
private let frameTypeUnary: UInt8 = 0
private let frameTypeStreamChunk: UInt8 = 1
private let frameTypeStreamEnd: UInt8 = 2
private let frameTypeError: UInt8 = 3
private let headerSize = 14

// MARK: - Frame

struct Frame {
    let version: UInt8
    let requestId: UInt32
    let frameType: UInt8
    let method: String
    let payload: Data
}

func encodeFrame(requestId: UInt32, frameType: UInt8, method: String, payload: Data) -> Data {
    let methodBytes = Array(method.utf8)
    let total = headerSize + methodBytes.count + payload.count
    var buffer = Data(capacity: total)

    buffer.append(protocolVersion)
    var rid = requestId.bigEndian
    buffer.append(Data(bytes: &rid, count: 4))
    buffer.append(frameType)
    var mlen = UInt32(methodBytes.count).bigEndian
    buffer.append(Data(bytes: &mlen, count: 4))
    buffer.append(contentsOf: methodBytes)
    var plen = UInt32(payload.count).bigEndian
    buffer.append(Data(bytes: &plen, count: 4))
    buffer.append(payload)

    return buffer
}

// MARK: - FrameReader

class FrameReader {
    private var buffer = Data()
    private var consumed = 0

    func add(_ chunk: Data) {
        buffer.append(chunk)
    }

    func tryReadFrame() -> Frame? {
        let available = buffer.count - consumed

        guard available >= headerSize else { return nil }

        let start = consumed
        let version = buffer[start]
        let requestId = buffer.readUInt32(at: start + 1)
        let frameType = buffer[start + 5]
        let methodLen = Int(buffer.readUInt32(at: start + 6))

        let neededForPayloadLen = headerSize + methodLen
        guard available >= neededForPayloadLen else { return nil }

        let payloadLen = Int(buffer.readUInt32(at: start + 10 + methodLen))
        let totalLen = neededForPayloadLen + payloadLen
        guard available >= totalLen else { return nil }

        let methodStart = start + 10
        let method = String(data: buffer[methodStart..<(methodStart + methodLen)], encoding: .utf8) ?? ""
        let payloadStart = start + neededForPayloadLen
        let payload = Data(buffer[payloadStart..<(payloadStart + payloadLen)])

        consumed += totalLen
        return Frame(version: version, requestId: requestId, frameType: frameType, method: method, payload: payload)
    }

    func compact() {
        guard consumed > 0 else { return }
        buffer = Data(buffer[consumed...])
        consumed = 0
    }
}

private extension Data {
    func readUInt32(at offset: Int) -> UInt32 {
        var value: UInt32 = 0
        _ = Swift.withUnsafeMutableBytes(of: &value) { dest in
            self.copyBytes(to: dest, from: offset..<(offset + 4))
        }
        return UInt32(bigEndian: value)
    }
}

// MARK: - Exception

struct KalamException: Error, CustomStringConvertible {
    let method: String
    let message: String

    var description: String { "KalamException(\(method)): \(message)" }
}

private func checkError(_ frame: Frame) throws {
    if frame.frameType == frameTypeError {
        throw KalamException(method: frame.method, message: String(data: frame.payload, encoding: .utf8) ?? "")
    }
}

// MARK: - Client

final class Kalam {
    static let shared = Kalam()
    private init() {}

    private var socketPath: String?
    private var fd: Int32 = -1
    private var nextRequestId: UInt32 = 1
    private let lock = NSLock()
    private let readQueue = DispatchQueue(label: "com.kalam.read", qos: .utility)
    private var pendingCalls: [UInt32: (Result<Frame, Error>) -> Void] = [:]
    private var pendingStreams: [UInt32: StreamCallbacks] = [:]
    private var reading = false

    private struct StreamCallbacks {
        let onChunk: (Frame) -> Void
        let onEnd: (Error?) -> Void
    }

    func useSockets(_ path: String) {
        disconnect()
        socketPath = path
    }

    private func connect() throws {
        guard let path = socketPath else {
            fatalError("Kalam not initialized. Call useSockets() first.")
        }

        let fd = socket(AF_UNIX, SOCK_STREAM, 0)
        guard fd >= 0 else { throw POSIXError(.init(rawValue: errno) ?? .ENOENT) }

        var addr = sockaddr_un()
        addr.sun_family = sa_family_t(AF_UNIX)
        let pathBytes = path.utf8CString
        withUnsafeMutablePointer(to: &addr.sun_path) { ptr in
            ptr.withMemoryRebound(to: CChar.self, capacity: 104) { dest in
                for (i, byte) in pathBytes.enumerated() {
                    dest[i] = byte
                }
            }
        }

        let addrLen = socklen_t(MemoryLayout<sockaddr_un>.size)
        let result = withUnsafePointer(to: &addr) { ptr in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockPtr in
                Darwin.connect(fd, sockPtr, addrLen)
            }
        }
        guard result == 0 else {
            Darwin.close(fd)
            throw POSIXError(.init(rawValue: errno) ?? .ECONNREFUSED)
        }

        self.fd = fd
        startReadLoop()
    }

    private func startReadLoop() {
        let fd = self.fd
        reading = true
        readQueue.async { [weak self] in
            let reader = FrameReader()
            let buf = UnsafeMutablePointer<UInt8>.allocate(capacity: 4096)
            defer { buf.deallocate() }

            while self?.reading == true {
                let n = Darwin.read(fd, buf, 4096)
                if n <= 0 { break }
                reader.add(Data(bytes: buf, count: n))

                while let frame = reader.tryReadFrame() {
                    reader.compact()
                    self?.dispatch(frame)
                }
            }
            self?.disconnect()
        }
    }

    private func dispatch(_ frame: Frame) {
        lock.lock()
        let id = frame.requestId

        if let completion = pendingCalls.removeValue(forKey: id) {
            lock.unlock()
            completion(.success(frame))
            return
        }

        if let callbacks = pendingStreams[id] {
            if frame.frameType == frameTypeStreamEnd {
                pendingStreams.removeValue(forKey: id)
                lock.unlock()
                callbacks.onEnd(nil)
            } else if frame.frameType == frameTypeError {
                pendingStreams.removeValue(forKey: id)
                lock.unlock()
                let err = KalamException(
                    method: frame.method,
                    message: String(data: frame.payload, encoding: .utf8) ?? ""
                )
                callbacks.onEnd(err)
            } else {
                lock.unlock()
                callbacks.onChunk(frame)
            }
            return
        }

        lock.unlock()
    }

    private func ensureConnected() throws {
        if fd < 0 { try connect() }
    }

    private func sendFrame(_ data: Data) {
        data.withUnsafeBytes { ptr in
            _ = Darwin.write(self.fd, ptr.baseAddress!, data.count)
        }
    }

    // MARK: Callback API (iOS 13+)

    func call(_ method: String, _ payload: Data, completion: @escaping (Result<Data, Error>) -> Void) {
        do {
            try ensureConnected()
        } catch {
            completion(.failure(error))
            return
        }

        lock.lock()
        let requestId = nextRequestId
        nextRequestId += 1
        lock.unlock()

        let frameData = encodeFrame(requestId: requestId, frameType: frameTypeUnary, method: method, payload: payload)

        lock.lock()
        pendingCalls[requestId] = { result in
            switch result {
            case .success(let frame):
                do {
                    try checkError(frame)
                    completion(.success(frame.payload))
                } catch {
                    completion(.failure(error))
                }
            case .failure(let error):
                completion(.failure(error))
            }
        }
        lock.unlock()
        sendFrame(frameData)
    }

    func stream(_ method: String, _ payload: Data, onChunk: @escaping (Data) -> Void, onEnd: @escaping (Error?) -> Void) {
        do {
            try ensureConnected()
        } catch {
            onEnd(error)
            return
        }

        lock.lock()
        let requestId = nextRequestId
        nextRequestId += 1
        lock.unlock()

        let frameData = encodeFrame(requestId: requestId, frameType: frameTypeUnary, method: method, payload: payload)

        lock.lock()
        pendingStreams[requestId] = StreamCallbacks(
            onChunk: { frame in
                do {
                    try checkError(frame)
                    onChunk(frame.payload)
                } catch {
                    onEnd(error)
                }
            },
            onEnd: onEnd
        )
        lock.unlock()
        sendFrame(frameData)
    }

    func disconnect() {
        reading = false
        if fd >= 0 {
            Darwin.close(fd)
            fd = -1
        }
        lock.lock()
        let calls = pendingCalls
        let streams = pendingStreams
        pendingCalls.removeAll()
        pendingStreams.removeAll()
        lock.unlock()

        for (_, completion) in calls {
            completion(.failure(KalamException(method: "", message: "Connection lost")))
        }
        for (_, callbacks) in streams {
            callbacks.onEnd(KalamException(method: "", message: "Connection lost"))
        }
    }
}

// MARK: - Server

protocol ServiceRouter {
    func handle(method: String, payload: Data, sink: ResponseSink)
}

final class ResponseSink {
    private let fd: Int32
    private let requestId: UInt32
    private let method: String

    init(fd: Int32, requestId: UInt32, method: String) {
        self.fd = fd
        self.requestId = requestId
        self.method = method
    }

    func sendUnary(_ payload: Data) {
        send(frameType: frameTypeUnary, payload: payload)
    }

    func sendChunk(_ payload: Data) {
        send(frameType: frameTypeStreamChunk, payload: payload)
    }

    func sendEnd() {
        send(frameType: frameTypeStreamEnd, payload: Data())
    }

    func sendError(_ message: String) {
        send(frameType: frameTypeError, payload: Data(message.utf8))
    }

    private func send(frameType: UInt8, payload: Data) {
        let data = encodeFrame(requestId: requestId, frameType: frameType, method: method, payload: payload)
        data.withUnsafeBytes { ptr in
            _ = Darwin.write(fd, ptr.baseAddress!, data.count)
        }
    }
}

final class KalamServer {
    static let shared = KalamServer()
    private init() {}

    private var serverFd: Int32 = -1
    private let acceptQueue = DispatchQueue(label: "com.kalam.accept", qos: .utility)
    private let clientQueue = DispatchQueue(label: "com.kalam.clients", qos: .utility, attributes: .concurrent)
    private var listening = false

    func serve(_ path: String, _ router: ServiceRouter) {
        unlink(path)

        let fd = socket(AF_UNIX, SOCK_STREAM, 0)
        guard fd >= 0 else { fatalError("Failed to create socket") }

        var addr = sockaddr_un()
        addr.sun_family = sa_family_t(AF_UNIX)
        let pathBytes = path.utf8CString
        withUnsafeMutablePointer(to: &addr.sun_path) { ptr in
            ptr.withMemoryRebound(to: CChar.self, capacity: 104) { dest in
                for (i, byte) in pathBytes.enumerated() {
                    dest[i] = byte
                }
            }
        }

        let addrLen = socklen_t(MemoryLayout<sockaddr_un>.size)
        _ = withUnsafePointer(to: &addr) { ptr in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockPtr in
                Darwin.bind(fd, sockPtr, addrLen)
            }
        }

        Darwin.listen(fd, 5)
        serverFd = fd
        listening = true

        print("Kalam server listening on \(path)")

        acceptQueue.async { [weak self] in
            while self?.listening == true {
                let clientFd = Darwin.accept(fd, nil, nil)
                if clientFd < 0 { break }
                self?.clientQueue.async {
                    self?.handleClient(clientFd, router)
                }
            }
        }
    }

    private func handleClient(_ clientFd: Int32, _ router: ServiceRouter) {
        let reader = FrameReader()
        let buf = UnsafeMutablePointer<UInt8>.allocate(capacity: 4096)
        defer {
            buf.deallocate()
            Darwin.close(clientFd)
        }

        while true {
            let n = Darwin.read(clientFd, buf, 4096)
            if n <= 0 { break }
            reader.add(Data(bytes: buf, count: n))

            while let request = reader.tryReadFrame() {
                reader.compact()
                let sink = ResponseSink(fd: clientFd, requestId: request.requestId, method: request.method)
                router.handle(method: request.method, payload: request.payload, sink: sink)
            }
        }
    }

    func close() {
        listening = false
        if serverFd >= 0 {
            Darwin.close(serverFd)
            serverFd = -1
        }
    }
}

File: runtime-dart/lib/kalam_runtime.dart
import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

const int _version = 1;
const int _frameTypeUnary = 0;
const int _frameTypeStreamChunk = 1;
const int _frameTypeStreamEnd = 2;
const int _frameTypeError = 3;

// Header: version(1) + requestId(4) + frameType(1) + methodLen(4) + payloadLen(4) = 14
const int _headerSize = 14;

// ── Frame ──────────────────────────────────────────────────────────────

class Frame {
  final int version;
  final int requestId;
  final int frameType;
  final String method;
  final Uint8List payload;

  Frame({
    required this.version,
    required this.requestId,
    required this.frameType,
    required this.method,
    required this.payload,
  });
}

Uint8List encodeFrame({
  required int requestId,
  required int frameType,
  required String method,
  required Uint8List payload,
}) {
  final methodBytes = Uint8List.fromList(method.codeUnits);
  final total = _headerSize + methodBytes.length + payload.length;
  final buffer = ByteData(total);
  var offset = 0;

  buffer.setUint8(offset, _version);
  offset += 1;
  buffer.setUint32(offset, requestId);
  offset += 4;
  buffer.setUint8(offset, frameType);
  offset += 1;
  buffer.setUint32(offset, methodBytes.length);
  offset += 4;

  final bytes = buffer.buffer.asUint8List();
  bytes.setRange(offset, offset + methodBytes.length, methodBytes);
  offset += methodBytes.length;

  buffer.setUint32(offset, payload.length);
  offset += 4;
  bytes.setRange(offset, offset + payload.length, payload);

  return bytes;
}

// ── FrameReader ────────────────────────────────────────────────────────

class FrameReader {
  final BytesBuilder _buffer = BytesBuilder(copy: false);
  int _consumed = 0;

  void add(List<int> chunk) {
    _buffer.add(chunk);
  }

  Frame? tryReadFrame() {
    final bytes = _buffer.toBytes();
    final available = bytes.length - _consumed;

    if (available < _headerSize) return null;

    final view = ByteData.sublistView(bytes, _consumed);

    final methodLen = view.getUint32(6);
    final neededForPayloadLen = _headerSize + methodLen;

    if (available < neededForPayloadLen) return null;

    final payloadLenOffset = 10 + methodLen;
    final payloadLen = view.getUint32(payloadLenOffset);
    final totalLen = neededForPayloadLen + payloadLen;

    if (available < totalLen) return null;

    final frameStart = _consumed;
    final frame = Frame(
      version: view.getUint8(0),
      requestId: view.getUint32(1),
      frameType: view.getUint8(5),
      method: String.fromCharCodes(bytes, frameStart + 10, frameStart + 10 + methodLen),
      payload: Uint8List.fromList(
        bytes.sublist(frameStart + neededForPayloadLen, frameStart + totalLen),
      ),
    );

    _consumed += totalLen;
    return frame;
  }

  void compact() {
    if (_consumed == 0) return;
    final bytes = _buffer.toBytes();
    final remaining = bytes.sublist(_consumed);
    _buffer.clear();
    _consumed = 0;
    if (remaining.isNotEmpty) {
      _buffer.add(remaining);
    }
  }
}

// ── Exception ──────────────────────────────────────────────────────────

class KalamException implements Exception {
  final String method;
  final String message;

  KalamException(this.method, this.message);

  @override
  String toString() => 'KalamException($method): $message';
}

void _checkError(Frame frame) {
  if (frame.frameType == _frameTypeError) {
    throw KalamException(
      frame.method,
      String.fromCharCodes(frame.payload),
    );
  }
}

// ── Client ─────────────────────────────────────────────────────────────

class Kalam {
  Kalam._();

  static final Kalam instance = Kalam._();

  String? _socketPath;
  Socket? _socket;
  FrameReader? _reader;
  int _nextRequestId = 1;
  final Map<int, Completer<Frame>> _pendingCalls = {};
  final Map<int, StreamController<Frame>> _pendingStreams = {};

  void useSockets(String path) {
    _socketPath = path;
    _disconnect();
  }

  Future<void> _ensureConnected() async {
    if (_socket != null) return;

    final address = InternetAddress(_socketPath!, type: InternetAddressType.unix);
    _socket = await Socket.connect(address, 0);
    _reader = FrameReader();

    _socket!.listen(
      _onData,
      onError: (_) => _disconnect(),
      onDone: _disconnect,
    );
  }

  void _onData(List<int> data) {
    _reader!.add(data);

    while (true) {
      final frame = _reader!.tryReadFrame();
      if (frame == null) break;
      _reader!.compact();

      final id = frame.requestId;

      // Route to pending unary call
      final completer = _pendingCalls.remove(id);
      if (completer != null) {
        completer.complete(frame);
        continue;
      }

      // Route to pending stream
      final controller = _pendingStreams[id];
      if (controller != null) {
        if (frame.frameType == _frameTypeStreamEnd) {
          _pendingStreams.remove(id);
          controller.close();
        } else if (frame.frameType == _frameTypeError) {
          _pendingStreams.remove(id);
          controller.addError(KalamException(
            frame.method,
            String.fromCharCodes(frame.payload),
          ));
          controller.close();
        } else {
          controller.add(frame);
        }
        continue;
      }
    }
  }

  void _disconnect() {
    _socket?.destroy();
    _socket = null;
    _reader = null;
    for (final c in _pendingCalls.values) {
      c.completeError(StateError('Connection lost'));
    }
    _pendingCalls.clear();
    for (final s in _pendingStreams.values) {
      s.addError(StateError('Connection lost'));
      s.close();
    }
    _pendingStreams.clear();
  }

  Future<Uint8List> call(String method, Uint8List payload) async {
    if (_socketPath == null) {
      throw StateError('Kalam not initialized. Call useSockets() first.');
    }

    await _ensureConnected();

    final requestId = _nextRequestId++;
    final completer = Completer<Frame>();
    _pendingCalls[requestId] = completer;

    _socket!.add(encodeFrame(
      requestId: requestId,
      frameType: _frameTypeUnary,
      method: method,
      payload: payload,
    ));

    final response = await completer.future;
    _checkError(response);
    return response.payload;
  }

  Stream<Uint8List> stream(String method, Uint8List payload) {
    if (_socketPath == null) {
      throw StateError('Kalam not initialized. Call useSockets() first.');
    }

    final requestId = _nextRequestId++;
    final controller = StreamController<Frame>();
    _pendingStreams[requestId] = controller;

    () async {
      await _ensureConnected();

      _socket!.add(encodeFrame(
        requestId: requestId,
        frameType: _frameTypeUnary,
        method: method,
        payload: payload,
      ));
    }();

    return controller.stream.map((frame) {
      _checkError(frame);
      return frame.payload;
    });
  }

  Future<void> close() async {
    _socket?.destroy();
    _socket = null;
    _reader = null;
    _pendingCalls.clear();
    _pendingStreams.clear();
  }
}

// ── Server ─────────────────────────────────────────────────────────────

abstract class ServiceRouter {
  Future<void> handle(String method, Uint8List payload, ResponseSink sink);
}

class ResponseSink {
  final Socket _client;
  final int _requestId;
  final String _method;

  ResponseSink(this._client, this._requestId, this._method);

  void sendUnary(Uint8List payload) {
    _client.add(encodeFrame(
      requestId: _requestId,
      frameType: _frameTypeUnary,
      method: _method,
      payload: payload,
    ));
  }

  void sendChunk(Uint8List payload) {
    _client.add(encodeFrame(
      requestId: _requestId,
      frameType: _frameTypeStreamChunk,
      method: _method,
      payload: payload,
    ));
  }

  void sendEnd() {
    _client.add(encodeFrame(
      requestId: _requestId,
      frameType: _frameTypeStreamEnd,
      method: _method,
      payload: Uint8List(0),
    ));
  }

  void sendError(String message) {
    _client.add(encodeFrame(
      requestId: _requestId,
      frameType: _frameTypeError,
      method: _method,
      payload: Uint8List.fromList(message.codeUnits),
    ));
  }
}

class KalamServer {
  KalamServer._();

  static final KalamServer instance = KalamServer._();

  ServerSocket? _server;

  Future<void> serve(String path, ServiceRouter router) async {
    final file = File(path);
    if (file.existsSync()) file.deleteSync();

    final address = InternetAddress(path, type: InternetAddressType.unix);
    _server = await ServerSocket.bind(address, 0);

    print('Kalam server listening on $path');

    await for (final client in _server!) {
      _handleClient(client, router);
    }
  }

  void _handleClient(Socket client, ServiceRouter router) {
    final reader = FrameReader();

    client.listen(
      (data) async {
        reader.add(data);

        while (true) {
          final request = reader.tryReadFrame();
          if (request == null) break;
          reader.compact();

          final sink = ResponseSink(client, request.requestId, request.method);
          try {
            await router.handle(request.method, request.payload, sink);
          } catch (e) {
            sink.sendError(e.toString());
          }
          await client.flush();
        }
      },
      onError: (e) => print('Client error: $e'),
      onDone: () => client.close(),
    );
  }

  Future<void> close() async {
    await _server?.close();
    _server = null;
  }
}
